ProperlyStopRunningPipelines.groovy

```diff
@@ -0,0 +1,102 @@
+/*
+Author: Alex Taylor
+Since: July 2018
+Description: This script stop all or a series of running pipeline jobs
+reset or became out of synch.
+Parameters: None
+Scope: Cloudbees Jenkins Platform
+*/
+
+/* This script can be used to kill off all running jobs, the latest build of a job, a specific job number, or jobs
+which have been running for a certain amount of time. This is all based on a couple of specific settings which are marked
+with comments. This script will guaranteed take 1.5 minutes to run because we want to ensure that each build command has the
+time needed to run*/
+
+import groovy.time.*;
+
+
+// A couple of example map definitions:
+
+//Used to kill all of the active jobs
+//def jobMap = [:]
+
+//Used to clean up specific builds. Put "buildNum: 0" if you want to stop the latest build
+/*def jobMap = [
+    1: [name:'$JOB_NAME', buildNum: $BUILD_NUM],
+    2: [name:'$JOB_NAME2', buildNum: $BUILD_NUM2],
+  	3: [name:'$JOB_NAME3', buildNum: $BUILD_NUM3]
+]
+*/
+
+def jobMap = [:]
+
+//Created for if the map is empty to add all the jobs
+if(jobMap.isEmpty()){
+  Jenkins.instance.getAllItems(org.jenkinsci.plugins.workflow.job.WorkflowJob).each{
+    job -> job.builds.findAll{it.isBuilding()}.each{
+      build ->
+      println("Adding: "+ job.fullName+ " build number " + build.getNumber().toInteger())
+      jobMap.put(++jobMap.size(), [ name: job.fullName, buildNum: build.getNumber().toInteger()])
+     }
+   }
+}
+use(TimeCategory)  {
+  def delay = 2.days;//Put in a Custom date here to kill anything older
+  // NO MODIFICATION
+  Jenkins.instance.getAllItems(org.jenkinsci.plugins.workflow.job.WorkflowJob).each{
+    job -> job.builds.findAll{it.isBuilding() && new Date(it.startTimeInMillis) < (new Date() - delay)}.each{
```

Comment:
Will cause a lot of unnecessary disk loading: http://javadoc.jenkins.io/hudson/util/RunList.html#byTimestamp-long-long- maybe?

Reply:
Right, if you force a read of every build in a system, this will take forever.